syscall/src/call.rs

```diff
@@ -0,0 +1,95 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+//
+// Copyright (c) 2024 Intel Corporation.
+//
+// Author: Chuanxiao Dong <chuanxiao.dong@intel.com>
+
+use super::*;
+use core::arch::asm;
+
+/// Macro to generate system call functions with varying numbers of arguments.
+#[macro_export]
+macro_rules! syscall {
+    ($($name:ident($a:ident, $($b:ident, $($c:ident, $($d:ident, $($e:ident, $($f:ident, )?)?)?)?)?);)+) => {
+        $(
+            /// Performs a system call with arguments.
+            ///
+            /// # Safety
+            ///
+            /// This function is unsafe because it performs a system call.
+            /// The kernel should check the syscall number and return the
+            /// expected result back.
+            #[allow(dead_code)]
+            pub unsafe fn $name($a: u64, $($b: u64, $($c: u64, $($d: u64, $($e: u64, $($f: u64)?)?)?)?)?) -> Result<u64, SysCallError> {
+                let mut ret = $a;
+                asm!(
+                    "int 0x80",
+                    inout("rax") ret,
+                    $(
+                        in("rdi") $b,
+                        $(
+                            in("rsi") $c,
+                            $(
+                                in("r8") $d,
+                                $(
+                                    in("r9") $e,
+                                    $(
+                                        in("r10") $f,
+                                    )?
+                                )?
+                            )?
+                        )?
+                    )?
+                    out("rcx") _,
+                    out("r11") _,
+                    options(nostack),
+                );
+
+                if ret > i32::MAX as u64 {
+                    // Convert to negative error code
+                    let error_code = ((ret as i64) - (1 << 32)) as i32;
+                    return Err(SysCallError::from(error_code))
+                }
```

Comment:
```suggestion
                if ret > (u64::MAX - u64::from(i32::MAX)) {
                    // Convert to negative error code
                    let error_code = ((ret as i64) - (1 << 32)) as i32;
                    return Err(SysCallError::from(error_code))
                }
```

The previous implementation classified `0x2_0000_0000` as an error, but that's not correct. 
We could also make the range even smaller (e.g. IIRC on Linux only everything from `0xffff_ffff_ffff_fe00` to `0xffff_ffff_ffff_ffff` is used for errors).

Reply:
Looks good to me.